### PR TITLE
Content security policy

### PR DESCRIPTION
**Pull request by snuggs**
# Content Security Policy _(MDN)_

### Reference Links:
  - https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP

### CODE REVIEW

**Comment by snuggs** [comment]
@brandondees  what did i clobber? didn't force push anything. But can't see what you wrote.

capture d ecran 2018-01-19 a 10 34 26

**Comment by brandondees** [comment]
it was just an approval. when the code changes, the review becomes obsolete

File: middleware/policy.es [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    header
+
+      // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
+      = 'Content-Security-Policy'
+
+      // Is this a security breach?
+      // Will someone be able to disable CSP with this?
+      // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only
+      + ( 'report' in context.request.query ? '-Report-Only' : '' )
```
Comment (snuggs) [inline middleware/policy.es]: @brandondees trying to figure out a way to trigger reporting. I feel having a querystring parameter `?report`
Should only work in development but that's a side effect that now has to be tracked. Apparently can do BOTH
`Content-Security-Policy` and `Content-Security-Policy-Report-Only` and they are complimentary and inconsequential. Perhaps that's our strategy. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only

Comment (brandondees) [reply middleware/policy.es]: wow am i behind the times on this. the report only feature looks like something everybody should enable all the time as a general best practice, but there are always exceptions to the rule for this kind of thing. where's this code going to have an effect? just on snuggsi.es? or for all users of the shim? or just on the docs website? i'm trying to gauge how much we're thinking about "show off and prescribe a best practice we figured out" vs "we need this to work right" vs "just exploring what's possible"

both csp and the reporting are things that there might be valid reasons to _not_ use, or to configure differently, naturally. but for typical web developers building apps, we should be able to come up with some sort of sensible default starting point that errs towards more secure rather than less. whitelist by default, blacklist instead only if it fits your model better, disable altogether if you really know what you're doing.

Comment (snuggs) [reply middleware/policy.es]: @brandondees great question. Basically for now eating own dogfood. (And great fallback reference of "we did this before here's where it's at since it's relevant moving forward". Great statements and thoughts. I'm having the same.

I found this as well. Perhaps `?report` just enables reporting. Should be inconsequential of the explicit `Content-Security-Policy`. And would only report on what is in `Content-Security-Policy-Report-Only`. I feel making them mirror images of each other ALWAYS would at least provide more Occam's razor and less Murphy's Law.

Thoughts...

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy#Multiple_content_security_policies

Comment (snuggs) [reply middleware/policy.es]: P.S. @brandondees better get up on it. YUGE part of Rails 5.2.

**Comment by snuggs** [comment]
capture d ecran 2018-01-21 a 04 22 43

@tmornini @aspleenic perfect example of a conditional *expression* vs a conditional *statement* (if).

**TL;DR** Statements & Expressions both need a predicate. But only one is used _(primarily)_ to return something. And the other is used _(primarily)_ for side effects.

Statements don't return values. Big epiphany for me recently was if statements aren't
_(or don't have to be)_ used for control flow at all. Merely used for side effects.
_(Unless you're programming in Ruby. But even Matz says that's a leaky abstraction)_.
The "control flow" is a bi-product of an if statement. A side effect of THE side effect if you will.

```ruby
"Hello World" if true # Leaky abstraction. Works in Ruby. Not natural to majority of other languages.
```
One thing that stands out is the previous statement cannot be composed. Expressions are composable
([as in this example within our content negotiation middleware](https://github.com/devpunks/snuggsi/blob/master/middleware/compressor.es#L34-L66).) I learned this in scheme/LISP programming where I had an "A-ha" moment where even the "if statements" were merely functions themselves. Yes they controlled flow. And yes they (not often) performed side effects (anti-pattern). But were "expressions" because they must return something or purity goes out the window re: flow in functional programming.

[#F Sharp / Haskell sheds more clarity and has great examples of the difference](https://fsharpforfunandprofit.com/posts/expressions-vs-statements/)
and [why imperative if statements and loops should be avoided](https://fsharpforfunandprofit.com/posts/control-flow-expressions) from a functional perspective. However a rubyist would consider everything above asinine because of that one glorious/dreadful bug/feature of an if statement returning a value. Ruby has lots of these bugs/features that we love to hate to love.

To be clear expressions and statements both control flow at the end of the day. I think we all agree on this. All the other stuff is what's debatable. ;-). I can say this though...IF statements are a cesspool for bugs. Another thing I think we all can agree on. [**...And the debate goes on** _(great song by the way)_](https://www.youtube.com/watch?v=slldMEPvUqA)

Feel free to chime in. Shared experience is always enlightening since control-flow effects every programmer ever. Cheers!

/cc @brandondees @btakita @kurtcagle @janz93 @mrbernnz 

image

**Comment by snuggs** [comment]
@brandondees P.S. `WIP` is off. As soon as we figure out what to do with `?report` we can merge. We can also do the reporting in a later PR. As that may neeed some insight on how/where the report should go. These are auto-magically `POST`ed from web browser to endpoint. How that endpoint logs I feel should be left open with a sane default.

Full test coverage ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅

**Comment by tmornini** [comment]
> Perhaps self is a better baseline. However, I prefer everything off in regards to security by default.

While I generally agree with your statement, I don’t believe it makes any sense to default to off if that means a thing is not allowed to access itself

**Comment by snuggs** [comment]
@brandondees @tmornini hmmmm in retrospect you bring up great points. Given some time to rethink what do you feel is the immediate feature to test. We have the (passing) tests that are spec approved. But TOO restrictive. Can't even get the server to show anything locally as is locked down like a fort. At first thought some config file but assumed we are smarter than that. @brandondees Define "Batteries included?" for CSP  (to the best of your recolleciton). What do we NOT want to think about?

**Comment by snuggs** [comment]
@brandondees @tmornini `policy.json` ? Would live quite nicely next to `policy.test` and `policy.es`

File: middleware/security.test [hunk 2]
```diff
@@ -8,20 +8,17 @@
```
Comment (snuggs) [inline middleware/security.test]: @brandondees @tmornini should this be called `policy` or `security`. The latter seems too vague in retrospect.

**Comment by brandondees** [comment]
as for what's "batteries included" for csp, I am not expert enough to really have a definitive answer to that. rails 5.2 added default CSP headers which might be a good place to look for some "sensible defaults" cues to follow. 

As a security wonk, I'd like to say lock it all down by default and require intentional whitelisting, but reality is that may be raising the friction level too high and would lead to folks disabling it altogether instead of making thoughtful and appropriate choices. It should probably let folks do the most basic low-risk things they expect out of the box (careful what you assume is low-risk on the web though), but provide some safety railing to keep folks out of the canyon.

**Comment by brandondees** [comment]
Happy to talk about the specific options individually if you want @snuggs
